luxonis_ml/nn_archive/config_building_blocks/base_models/head.py

```diff
@@ -190,93 +148,146 @@ def validate_label_type(
         return value
 
 
-class HeadInstanceSegmentationYOLO(HeadObjectDetectionYOLO, HeadSegmentation, ABC):
-    """Metadata for YOLO instance segmentation head.
+class HeadObjectDetectionYOLO(HeadObjectDetection, HeadSegmentation, ABC):
+    """Metadata for YOLO object detection head.
 
     @type family: str
     @ivar family: Decoding family.
-    @type outputs: C{OutputsInstanceSegmentationYOLO}
+    @type outputs: Union[C{OutputsYOLO}, C{OutputsInstanceSegmentationYOLO}, C{OutputsKeypointDetectionYOLO}, C{OutputsOBBDetectionYOLO}]
     @ivar outputs: A configuration specifying which output names from the `outputs` block of the archive are fed into the head.
-    @type postprocessor_path: str
-    @ivar postprocessor_path: Path to the secondary executable used in YOLO instance
-        segmentation.
-    @type n_prototypes: int
-    @ivar n_prototypes: Number of prototypes per bbox.
+    @type subtype: ObjectDetectionSubtypeYOLO
+    @ivar subtype: YOLO family decoding subtype (e.g. v5, v6, v7 etc.).
+    @type postprocessor_path: str | None
+    @ivar postprocessor_path: Path to the secondary executable used in YOLO instance segmentation.
+    @type n_prototypes: int | None
+    @ivar n_prototypes: Number of prototypes per bbox in YOLO instance segmnetation.
+    @type n_keypoints: int | None
+    @ivar n_keypoints: Number of keypoints per bbox in YOLO keypoint detection.
+    @type is_softmax: bool | None
+    @ivar is_softmax: True, if output is already softmaxed in YOLO instance segmentation.
     """
 
-    family: Literal["InstanceSegmentationYOLO"] = Field(
-        ..., description="Decoding family."
-    )
-    outputs: OutputsInstanceSegmentationYOLO = Field(
+    family: Literal[
```

Comment:
Do those to actually need to be explicitly defined? I think we limit ourselves to only one type of output, whereas we could have family be just "YOLO" or "DetectionYOLO" and output some general OutputsYOLO which can have additional fields on the output?

Reply:
Okay, I made a generic YOLO head with optional parameters and a generic Yolo output with optional parameters. Validation is performed based on the defined attributes and then the Output attributes are also validated to match the specific task. This introduces more validation complexity. Added in [e6ddc8e](https://github.com/luxonis/luxonis-ml/pull/112/commits/e6ddc8ec4d9199bb47d82535b71fe6fba3c3b6cf).